### PR TITLE
Better articulate server competition challenges

### LINKED ISSUE #1794: clearer advice on server interpretation
In rfc7540, the 'exclusive' priority description gave explicity advice on when it needs to be disregarded:

```
rfc7540 5.3.1
 ...
 Inside the dependency tree, a dependent stream SHOULD only be
 allocated resources if either all of the streams that it depends on
 (the chain of parent streams up to 0x0) are closed or it is not
 possible to make progress on them."
```

I feel the new priorities draft is not explicit enough on this and implemenations might misinterpret its requirements as in

```
draft: 10.
 Therefore, non-incremental responses of the same urgency SHOULD 
  be served in their entirety, one-by-one, based on the stream ID[...]
```

I propose to change:

```
draft 10.
 It is RECOMMENDED that, when possible, servers respect the 
urgency parameter (Section 4.1), sending higher urgency  responses 
before lower urgency responses.
```

to

```
 Connections SHOULD be progressed whenever response data is available 
 and urgency/incremental information MUST not prevent use of downstream 
 bandwidth. Given that, it is RECOMMENDED that servers respect the urgency 
 parameter (Section 4.1), sending higher urgency  responses before lower 
 urgency responses.
```

### PR DESCRIPTION
**Pull request by LPardue**
closes #1794

This is a derivative of my suggestion on https://github.com/httpwg/http-extensions/issues/1794#issuecomment-974212727

### CODE REVIEW

File: draft-ietf-httpbis-priority.md [hunk 1]
```diff
@@ -666,11 +666,15 @@ about the requests it initiates.
...
-Priority signals are input to a prioritization process. They do not guarantee
-any particular processing or transmission order for one response relative to any
-other response. An endpoint cannot force a peer to process concurrent request in
-a particular order using priority. Expressing priority is therefore only a
-suggestion.
+Server scheduling is a prioritization process, based on many inputs and factors,
+that determine how connection resources such as bandwidth are used. This section
+describes considerations that servers can make when scheduling.
+
+Priority signals are only one form of input to a prioritization process. They do
```
Comment (kazuho) [inline draft-ietf-httpbis-priority.md:673]: How about something like:
```suggestion
It is generally beneficial for an HTTP server to send all responses as early as
possible. However, when serving multiple requests on a single connection, there
could be competition between the requests for resources such as connection
bandwidth. This section describes considerations regarding how servers can
schedule the order in which the competing responses will be sent, when such
competition exists.

Server scheduling is a prioritization process based on many inputs and factors,
with priority signals being only one form of input. Priority signals do
```

Regarding #1794, I think there are two paths forward. One is to explain how responses should be served in general, and the other is to limit the scope of this section to how to prioritize one above another when there is competition.

The change proposed in this comment takes the latter approach, while pointing _out as a fact_ that generally speaking serving every response ASAP is beneficial (hence subsuming the new paragraph being proposed below).

File: draft-ietf-httpbis-priority.md [hunk 2]
```diff
@@ -682,6 +686,14 @@ basic recommendations for implementations.
 Clients cannot depend on particular treatment based on priority signals. Servers
 can use other information to prioritize responses.
 
+It is generally beneficial for the server to maximize its use of available
+bandwidth when serving concurrent requests. Bandwidth utilization for any given
+connection is an output of scheduling that can be used in a feedback loop for
+subsequent scheduling decisions. For example, a server that prioritizes sending
+response data for a resource that it can only read slowly might find bandwidth
+is under-utilized. The server could redistribute the available bandwidth to
+other requests.
```
Comment (kazuho) [inline draft-ietf-httpbis-priority.md:695]: I think we can drop this paragraph If we adopt something similar to what I have proposed above.

I get the intent of this addition, but to me it is a bit hard to digest. I have difficulty understanding the second sentence. Also, I'm not sure if "re-"distribute found in the third sentence is precise. The text on `urgency` as a suggesting the precedence between requests, therefore, "re-"distribution should be unnecessary.

Another argument against this clause is that HTTP servers are required to serve responses in timely manner regardless of the existence of a prioritization mechanism and therefore that this type of recommendation is not strictly necessary.

**Review by kazuho — COMMENTED** [review]
Thank you for the PR. I agree that we need to address this. I like the first part of the change, but the second part is difficult for me to digest.

File: draft-ietf-httpbis-priority.md (hunk 1 above)
Comment (LPardue) [reply draft-ietf-httpbis-priority.md:673]: I like this!

File: draft-ietf-httpbis-priority.md (hunk 2 above)
Comment (LPardue) [reply draft-ietf-httpbis-priority.md]: removed the para

**Review by kazuho — COMMENTED** [review]
@LPardue Thank you for the changes.

I think this PR by itself might be ready for merge, though I see some amount of redundant text in the following paragraphs.

PR #1822 is my attempt to address this redundancy. It builds on top of this PR, merging main and then eliminating redundancy in commit 8be14c9.

PTAL.

**Comment by LPardue** [comment]
We've resolved the comments and merged the related PR. I'm taking it as implicit approval and merging.

**Pull request merged**